### ISSUE TITLE
NDESC defaults to 2^32-1 in ToDPDKDevice

### ISSUE THREAD

**Issue by davidek** [tbarbette/fastclick#9]
It appears that the last commits introduced a small issue with the default `NDESC` in `ToDPDKDevice`: if the value is not specified at configuration time, then `QueueDevice::ndesc` (which is an `int`) gets `-1` ([here](https://github.com/tbarbette/fastclick/blob/master%40%7B2016-02-24%7D/elements/userlevel/queuedevice.cc#L31)), but then it is passed to `DPDKDevice::add_tx_device` ([here](https://github.com/tbarbette/fastclick/blob/master%40%7B2016-02-24%7D/elements/userlevel/todpdkdevice.cc#L85)), which takes an `unsigned`. Also related, in the ternary expression [here](https://github.com/tbarbette/fastclick/blob/master%40%7B2016-02-24%7D/lib/dpdkdevice.cc#L160) and [here](https://github.com/tbarbette/fastclick/blob/master%40%7B2016-02-24%7D/lib/dpdkdevice.cc#L168) the condition is always true (because it's unsigned). 

Of course, this causes the call to `rte_eth_tx_queue_setup` to fail.
Trivial workaround is to manually specify `NDESC` when configuring `ToDPDKDevice`.

**Comment by tbarbette** [comment]
Solved?

**Comment by davidek** [comment]
It appears that the specific issue with `ToDPDKDevice` is gone, but it's still very unhappy when `NDESC` is not defined -- now also in the `FromDPDKDevice` element: now, instead, it appears that when NDESC is unspecified it takes `0` and DPDK doesn't like it. Perhaps I'm missing something and you indented `NDESC` to be required in the conf file?

Reproducing:
```
./configure --enable-multithread --disable-linuxmodule --enable-intel-cpu --enable-user-multithread --verbose CFLAGS="-g -O0" CXXFLAGS="-g -std=gnu++11 -O0" --disable-dynamic-linking --enable-poll --enable-bound-port-transfer --with-netmap=no --enable-zerocopy --disable-dpdk-pools --enable-dpdk --disable-batch
```
conf file: `FromDPDKDevice(0)  -> ToDPDKDevice(0);`
run : `sudo userlevel/click -c 0x1 -n 1 -- -f ...`

Some `printf`ing (`git apply`-able)
```diff
diff --git a/lib/dpdkdevice.cc b/lib/dpdkdevice.cc
index 4a8c1c9..e31964d 100644
--- a/lib/dpdkdevice.cc
+++ b/lib/dpdkdevice.cc
@@ -177,7 +177,10 @@ int DPDKDevice::initialize_device(unsigned port_id, DevInfo &info,
                 port_id, i, info.n_rx_descs, numa_node, &rx_conf,
                 _pktmbuf_pools[numa_node]) != 0)
+        {
+            fprintf(stderr, "info.n_rx_descs: %d\n", info.n_rx_descs);
             return errh->error(
                 "Cannot setup RX queue %u of port %u on node %u",
                 i, port_id, numa_node);
+        }
     }
 
@@ -185,7 +188,10 @@ int DPDKDevice::initialize_device(unsigned port_id, DevInfo &info,
         if (rte_eth_tx_queue_setup(port_id, i, info.n_tx_descs, numa_node,
                                    &tx_conf) != 0)
+        {
+            fprintf(stderr, "info.n_tx_descs: %d\n", info.n_tx_descs);
             return errh->error(
                 "Cannot setup TX queue %u of port %u on node %u",
                 i, port_id, numa_node);
+        }
 
     int err = rte_eth_dev_start(port_id);
```

Output:
```
Initializing DPDK
info.n_rx_descs: 0
/home/davide/master/click_conf/bug_copy.click:3: While initializing 'ToDPDKDevice@2 :: ToDPDKDevice':
  Cannot setup RX queue 0 of port 0 on node 0
Router could not be initialized!
```
And, when running with `FromDPDKDevice(0, NDESC 1024) -> ToDPDKDevice(0);`:
```
Initializing DPDK
PMD: ixgbe_dev_rx_queue_setup(): sw_ring=0x7efdf669d140 sw_sc_ring=0x7efdf669b000 hw_ring=0x7efdf669f280 dma_addr=0x83669f280
info.n_tx_descs: 0
/home/davide/master/click_conf/bug_copy.click:3: While initializing 'ToDPDKDevice@2 :: ToDPDKDevice':
  Cannot setup TX queue 0 of port 0 on node 0
Router could not be initialized!
```

**Comment by tbarbette** [comment]
Solved by 117e01759cac44b79eb5e47c871cb584e4a91fa0 , the default value is set in the To/FromDPDK device in mainline click but wasn't set there in fastclick ,so retrieving the patch as is lead to loose the default value.

Fixed, thanks again !
